Adding Arcus.WebApi.OpenApi.Extensions project (#46)
arcus-azure/arcus.webapi: docs/features/openapi.securitydefinitions.md

---
title: "Adding security information to OpenAPI documentation"
layout: default
---

## Adding OAuth security definition to API operations

### Introduction

> When an API is secured via OAuth, it is helpful if the Open API documentation makes this clear via a security scheme and the API operations that require authorization automatically inform the consumer that it is possible that a 401 (Unauthorized) or 409 (Forbidden) response is returned.

Reviewer: ```suggestion
When an API is secured via OAuth, it is helpful if the Open API documentation makes this clear via a security scheme and the API operations that require authorization automatically inform the consumer that it is possible that a 401 Unauthorized or 403 Forbidden response is returned.
```

> The `OAuthAuthorizeOperationFilter` that is part of this package exposes this functionality.

Reviewer: I'd go with something along those lines:
```suggestion
`OAuthAuthorizeOperationFilter` will enrich your OpenAPI documentation with the relevant information.
```

### Usage

> To indicate that an API is protected by OAuth, and enable Swagger UI to retrieve a JWT token when invoking API operations via Swagger, add the following code to the `ConfigureServices` method of the `Startup` class:

Reviewer: ```suggestion
To indicate that an API is protected by OAuth, you need to add `AuthorizeCheckOperationFilter` as an `OperationFilter` when configuring Swashbuckles Swagger generation:
```

Reviewer: I would move the following more to the introduction:
> and enable Swagger UI to retrieve a JWT token when invoking API operations via Swagger

```csharp

> services.AddSwaggerGen(c =>

Reviewer: This is documentation so not a blocker for me, but think indeed that it would improve readability

> services.AddSwaggerGen(c =>

Author: Came up with something better. :)

{
   c.SwaggerDoc("v1", new Info { Title = "My API v1", Version = "v1" });

   c.AddSecurityDefinition("oauth2", new OAuth2Scheme
   {
      Flow = "implicit",
      AuthorizationUrl = $"{authorityUrl}connect/authorize",

>       Scopes = scopes

Reviewer 2: Maybe an example on the `scopes`?

Reviewer: Actually it's passed to the filter already but not used here so they need to be aligned indeed

   });

>    c.OperationFilter<OAuthAuthorizeOperationFilter>("myApiScope1", "myApiScope2");

Reviewer 2: These scopes are passed allong the filter in a way you would think the parameters are decorated with a `params` syntactic sugar; but the input is an `IEnumerable<_>`; not sure though if this is correct?

Author: True, fixed this.  Thx.

>    c.OperationFilter<AuthorizeCheckOperationFilter>("myApiScope1", "myApiScope2");

Reviewer 2: I suspect that you should use our operation filter here instead of the `AuthorizeCheckOperationFilter` from a [demo project](https://github.com/dvanherten/swagger-identityserver/blob/master/WebApi/AuthorizeCheckOperationFilter.cs). :)

Reviewer: What do you mean with "our operation filter"? Are you referring to the shared access signature + certificate filter?

If so, we've discussed this and will exclude it in the first iteration.

Reviewer 2: The operation filter in the PR :)

Reviewer: Ah I see, naming is not inline, sorry about that :)

Author: Should be fixed.

Reviewer: Thanks!
